src/constraint.c

```diff
@@ -109,7 +109,8 @@ ts_constraint_find_matching(HeapTuple ht_tup, Relation chunk_rel)
 		{
 			if (ts_indexing_compare(ht_con->conindid, chunk_con->conindid))
 			{
-				result = chunk_con;
+				result = (Form_pg_constraint) palloc(sizeof(FormData_pg_constraint));
+				memcpy(result, chunk_con, sizeof(FormData_pg_constraint));
```

Comment:
Maybe also add a note in the comment that it is up to caller to pfree the constraint form or manage it via an appropriate memory context.